File: src/main/java/com/epri/dss/general/impl/GrowthShapeObjImpl.java
package com.epri.dss.general.impl;

import java.io.PrintStream;

import com.epri.dss.common.DSSClass;
import com.epri.dss.common.impl.Utilities;
import com.epri.dss.general.GrowthShape;
import com.epri.dss.general.GrowthShapeObj;

public class GrowthShapeObjImpl extends DSSObjectImpl implements GrowthShapeObj {

	/* Number of points in curve */
	private int Npts;
	/* Number of years presently allocated in look up table */
	private int NYears;
	private int BaseYear;

	/* Year values */
	private int[] Year;
	private double[] YearMult, Multiplier;

	public GrowthShapeObjImpl(DSSClass ParClass, String GrowthShapeName) {
		super(ParClass);

		setName(GrowthShapeName.toLowerCase());
		this.DSSObjType = ParClass.getDSSClassType();

		this.Npts = 0;
		this.Year = null;
		this.Multiplier = null;
		this.NYears = 30;
		this.YearMult = new double[this.NYears];

		initPropertyValues(0);
	}

	/**
	 * Get multiplier for specified year.
	 *
	 * This function returns the multiplier to use for a load in the given year.
	 * The first year specified in the curve is the base year.  The base value
	 * is the beginning of the first year.
	 */
	public double getMult(int Yr) {
		int Index;

		double Result = 1.0;  // default return value if no points in curve

		if (Npts > 0) {  // Handle Exceptional cases
			Index = Yr - BaseYear;
			if (Index > 0) {  // Returns 1.0 for base year or any year previous  TODO Check zero based indexing

				if (Index > NYears) {  // Make some more space
					NYears = Index + 10;
					YearMult = (double[]) Utilities.resizeArray(YearMult, NYears);
					reCalcYearMult();
				}

				Result = YearMult[Index];

			}

		}

		return Result;
	}

	/* FIXME Private procedure in OpenDSS */
	public void reCalcYearMult() {
		// Fill up the YearMult array with total yearly multiplier from base year
		double Mult = Multiplier[1];
		double MultInc = Mult;
		YearMult[0] = Mult;  // TODO Check zero based indexing
		int DataPtr = 1;  // TODO Check zero based indexing
		int Yr = BaseYear;
		for (int i = 1; i < NYears; i++) {  // TODO Check zero based indexing
			Yr += 1;
			if (DataPtr < Npts) {
				if (Year[DataPtr + 1] == Yr) {
					DataPtr += 1;
					MultInc = Multiplier[DataPtr];
				}
			}
			Mult = Mult * MultInc;
			YearMult[i] = Mult;
		}
	}

	public void dumpProperties(PrintStream F, boolean Complete) {
		super.dumpProperties(F, Complete);

		for (int i = 0; i < ParentClass.getNumProperties(); i++) {
			switch (i) {
			case 2:  // TODO Check zero based indexing
				F.println("~ " + ParentClass.getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			case 3:
				F.println("~ " + ParentClass.getPropertyName()[i] + "=(" + getPropertyValue(i) + ")");
				break;
			default:
				F.println("~ " + ParentClass.getPropertyName()[i] + "=" + getPropertyValue(i));
				break;
			}
		}
	}

	public String getPropertyValue(int Index) {
		int i;
		String Result;

		switch (Index) {
		case 2:  // TODO Check zero based indexing
			Result = "(";
			break;
		case 3:
			Result = "(";
			break;
		default:
			Result = "";
			break;
		}

		switch (Index) {
		case 2:
			for (i = 0; i < Npts; i++)
				Result = Result + String.format("%-d, ", Year[i]);
			break;
		case 3:
			for (i = 0; i < Npts; i++)
				Result = Result + String.format("%-g, ", Multiplier[i]);
			break;
		default:
			Result = super.getPropertyValue(Index);
			break;
		}

		switch (Index) {
		case 2:
			Result = Result + ")";
			break;
		case 3:
			Result = Result + ")";
			break;
		}

		return Result;
	}

	public void initPropertyValues(int ArrayOffset) {

		PropertyValue[0] = "0";  // Number of points to expect
		PropertyValue[1] = "";   // vector of year values
		PropertyValue[2] = "";   // vector of multiplier values corresponding to years
		PropertyValue[3] = "";   // Switch input to a csvfile                (year, mult)
		PropertyValue[4] = "";   // switch input to a binary file of singles (year, mult)
		PropertyValue[5] = "";   // switch input to a binary file of doubles (year, mult)

		super.initPropertyValues(GrowthShape.NumPropsThisClass);
	}


	/* FIXME Private members in OpenDSS */
	public int getNpts() {
		return Npts;
	}

	public void setNpts(int npts) {
		Npts = npts;
	}

	public int[] getYear() {
		return Year;
	}

	public void setYear(int[] year) {
		Year = year;
	}

	public double[] getMultiplier() {
		return Multiplier;
	}

	public void setMultiplier(double[] multiplier) {
		Multiplier = multiplier;
	}

	public int getBaseYear() {
		return BaseYear;
	}

	public void setBaseYear(int baseYear) {
		BaseYear = baseYear;
	}

}

File: pydss/general/GrowthShapeObjImpl.py
from dss.general.impl.DSSObjectImpl import (DSSObjectImpl,)
from dss.common.impl.Utilities import (Utilities,)
from dss.general.GrowthShapeObj import (GrowthShapeObj,)
from dss.general.GrowthShape import (GrowthShape,)


class GrowthShapeObjImpl(DSSObjectImpl, GrowthShapeObj):
    npts = None
    nYears = None
    baseYear = None
    year = None
    yearMult = None
    multiplier = None

    def __init__(self, parClass, growthShapeName):
        super(GrowthShapeObjImpl, self)(parClass)
        self.setName(growthShapeName.toLowerCase())
        self.objType = parClass.getDSSClassType()
        self.npts = 0
        self.year = None
        self.multiplier = None
        self.nYears = 30
        self.yearMult = [None] * self.nYears
        self.initPropertyValues(0)

    def getMult(self, yr):
        """Get multiplier for specified year.

        This function returns the multiplier to use for a load in the given year.
        The first year specified in the curve is the base year.  The base value
        is the beginning of the first year.
        """
        # FIXME Private procedure in OpenDSS
        result = 1.0
        # default return value if no points in curve
        if self.npts > 0:
            # handle exceptional cases
            index = yr - self.baseYear
            if index > 0:
                # returns 1.0 for base year or any year previous  TODO Check zero based indexing
                if index > self.nYears:
                    # make some more space
                    self.nYears = index + 10
                    self.yearMult = Utilities.resizeArray(self.yearMult, self.nYears)
                    self.reCalcYearMult()
                result = self.yearMult[index]
        return result

    def reCalcYearMult(self):
        # fill up the yearMult array with total yearly multiplier from base year
        mult = self.multiplier[1]
        multInc = mult
        self.yearMult[0] = mult
        # TODO Check zero based indexing
        dataPtr = 1
        # TODO Check zero based indexing
        yr = self.baseYear
        _0 = True
        i = 1
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.nYears):
                break
            # TODO Check zero based indexing
            yr += 1
            if dataPtr < self.npts:
                if self.year[dataPtr + 1] == yr:
                    dataPtr += 1
                    multInc = self.multiplier[dataPtr]
            mult = mult * multInc
            self.yearMult[i] = mult

    def dumpProperties(self, f, complete):
        super(GrowthShapeObjImpl, self).dumpProperties(f, complete)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.parentClass.getNumProperties()):
                break
            _1 = i
            _2 = False
            while True:
                if _1 == 2:
                    _2 = True
                    print '~ ' + self.parentClass.getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if (_2 is True) or (_1 == 3):
                    _2 = True
                    print '~ ' + self.parentClass.getPropertyName()[i] + '=(' + self.getPropertyValue(i) + ')'
                    break
                if True:
                    _2 = True
                    print '~ ' + self.parentClass.getPropertyName()[i] + '=' + self.getPropertyValue(i)
                    break
                break

    def getPropertyValue(self, index):
        _0 = index
        _1 = False
        while True:
            if _0 == 2:
                _1 = True
                result = '('
                break
            if (_1 is True) or (_0 == 3):
                _1 = True
                result = '('
                break
            if True:
                _1 = True
                result = ''
                break
            break
        _2 = index
        _3 = False
        while True:
            if _2 == 2:
                _3 = True
                _4 = True
                i = 0
                while True:
                    if _4 is True:
                        _4 = False
                    else:
                        i += 1
                    if not (i < self.npts):
                        break
                    result = result + String.format.format('%-d, ', self.year[i])
                break
            if (_3 is True) or (_2 == 3):
                _3 = True
                _5 = True
                i = 0
                while True:
                    if _5 is True:
                        _5 = False
                    else:
                        i += 1
                    if not (i < self.npts):
                        break
                    result = result + String.format.format('%-g, ', self.multiplier[i])
                break
            if True:
                _3 = True
                result = super(GrowthShapeObjImpl, self).getPropertyValue(index)
                break
            break
        _6 = index
        _7 = False
        while True:
            if _6 == 2:
                _7 = True
                result = result + ')'
                break
            if (_7 is True) or (_6 == 3):
                _7 = True
                result = result + ')'
                break
            break
        return result

    def initPropertyValues(self, arrayOffset):
        # FIXME Private members in OpenDSS
        self.propertyValue[0] = '0'
        # number of points to expect
        self.propertyValue[1] = ''
        # vector of year values
        self.propertyValue[2] = ''
        # vector of multiplier values corresponding to years
        self.propertyValue[3] = ''
        # switch input to a csvfile                (year, mult)
        self.propertyValue[4] = ''
        # switch input to a binary file of singles (year, mult)
        self.propertyValue[5] = ''
        # switch input to a binary file of doubles (year, mult)
        super(GrowthShapeObjImpl, self).initPropertyValues(GrowthShape.NumPropsThisClass)

    def getNPts(self):
        return self.npts

    def setNPts(self, n):
        self.npts = n

    def getYear(self):
        return self.year

    def setYear(self, values):
        self.year = values

    def getMultiplier(self):
        return self.multiplier

    def setMultiplier(self, mult):
        self.multiplier = mult

    def getBaseYear(self):
        return self.baseYear

    def setBaseYear(self, year):
        self.baseYear = year
